geode-core/src/distributedTest/java/org/apache/geode/internal/cache/execute/FunctionOnRegionRetryDUnitTest.java

```diff
@@ -0,0 +1,130 @@
+
+      "HA | CLIENT_HAS_METADATA | OBJECT_REFERENCE | -1 | 9",
+      "HA | CLIENT_HAS_METADATA | OBJECT_REFERENCE | 0 | 3",
+      "HA | CLIENT_HAS_METADATA | OBJECT_REFERENCE | 2 | 9",
+      "HA | CLIENT_HAS_METADATA | STRING | -1 | 9",
+      "HA | CLIENT_HAS_METADATA | STRING | 0 | 3",
+      "HA | CLIENT_HAS_METADATA | STRING | 2 | 9",
+  })
+  @TestCaseName("[{index}] {method}: {params}")
+  public void testOnRegion(final HAStatus haStatus,
+      final ClientMetadataStatus clientMetadataStatus,
+      final FunctionIdentifierType functionIdentifierType,
+      final int retryAttempts,
+      final int expectedCalls) throws Exception {
+
+    Function function = testFunctionRetry(haStatus,
+        clientMetadataStatus,
+        ExecutionTarget.REGION,
+        functionIdentifierType,
+        retryAttempts);
+
+    GeodeAwaitility.await("Awaiting getNumberOfFunctionCalls isEqualTo expectedCalls")
+        .atMost(30, TimeUnit.SECONDS)
+        .untilAsserted(
+            () -> assertThat(getNumberOfFunctionCalls(function.getId())).isEqualTo(expectedCalls));
+
+  }
+
+
+  @Test
+  @Parameters({
+      /*
+       * haStatus | clientMetadataStatus | functionIdentifierType | retryAttempts | expectedCalls
+       */
+      "NOT_HA | CLIENT_MISSING_METADATA | OBJECT_REFERENCE | -1 | 1",
+      "NOT_HA | CLIENT_MISSING_METADATA | OBJECT_REFERENCE | 0 | 1",
+      "NOT_HA | CLIENT_MISSING_METADATA | OBJECT_REFERENCE | 2 | 1",
+      "NOT_HA | CLIENT_MISSING_METADATA | STRING | -1 | 1",
+      "NOT_HA | CLIENT_MISSING_METADATA | STRING | 0 | 1",
+      "NOT_HA | CLIENT_MISSING_METADATA | STRING | 2 | 1",
+      "HA | CLIENT_MISSING_METADATA | OBJECT_REFERENCE | -1 | 3",
+      "HA | CLIENT_MISSING_METADATA | OBJECT_REFERENCE | 0 | 1",
+      "HA | CLIENT_MISSING_METADATA | OBJECT_REFERENCE | 2 | 3",
+      "HA | CLIENT_MISSING_METADATA | STRING | -1 | 3",
+      "HA | CLIENT_MISSING_METADATA | STRING | 0 | 1",
+      "HA | CLIENT_MISSING_METADATA | STRING | 2 | 3",
+
+      "HA | CLIENT_HAS_METADATA | OBJECT_REFERENCE | -1 | 3",
+      "HA | CLIENT_HAS_METADATA | OBJECT_REFERENCE | 0 | 1",
+      "HA | CLIENT_HAS_METADATA | OBJECT_REFERENCE | 2 | 3",
+      "HA | CLIENT_HAS_METADATA | STRING | -1 | 3",
+      "HA | CLIENT_HAS_METADATA  | STRING | 3 | 4",
+  })
+  @TestCaseName("[{index}] {method}: {params}")
+  public void testOnRegionWithSingleKeyFilter(final HAStatus haStatus,
+      final ClientMetadataStatus clientMetadataStatus,
+      final FunctionIdentifierType functionIdentifierType,
+      final int retryAttempts,
+      final int expectedCalls) throws Exception {
+
+    Function function = testFunctionRetry(haStatus,
+        clientMetadataStatus,
+        ExecutionTarget.REGION_WITH_FILTER_1_KEY,
+        functionIdentifierType,
+        retryAttempts);
+
+    GeodeAwaitility.await("Awaiting getNumberOfFunctionCalls isEqualTo expectedCalls")
+        .atMost(30, TimeUnit.SECONDS)
```

Comment:
I think we're supposed to just use the default GeodeAwaitility timeout in case the test is executing during a stop-the-world GC pause or is running on a machine with super busy CPUs.

Reply:
@kirklund Thanks. Pushed changes to use default timeout with GeodeAwaitility.